File: DEVELOPMENT/Python/gd_riccardo/C translation/ransac_fit.c
#include <stdio.h>
#include <string.h>
#include <stdint.h>
#include <math.h>

#include "ransac_fit.h"

/*
 * Solve a 3x3 linear system Ax = b using Cramer's Rule.
 * A is stored row-major: A[row][col]
 * Returns 1 on success, 0 if the matrix is singular.
 * 
 */
static int solve_3x3(double A[3][3], double b[3], double x[3]) {
    double det = A[0][0] * (A[1][1] * A[2][2] - A[1][2] * A[2][1])
               - A[0][1] * (A[1][0] * A[2][2] - A[1][2] * A[2][0])
               + A[0][2] * (A[1][0] * A[2][1] - A[1][1] * A[2][0]);

    if (fabs(det) < 1e-10)
        return 0;

    double inv_det = 1.0 / det;

    double det0 = b[0]    * (A[1][1] * A[2][2] - A[1][2] * A[2][1])
                - A[0][1] * (b[1]    * A[2][2] - A[1][2] * b[2]   )
                + A[0][2] * (b[1]    * A[2][1] - A[1][1] * b[2]   );

    double det1 = A[0][0] * (b[1]    * A[2][2] - A[1][2] * b[2]   )
                - b[0]    * (A[1][0] * A[2][2] - A[1][2] * A[2][0])
                + A[0][2] * (A[1][0] * b[2]    - b[1]    * A[2][0]);

    double det2 = A[0][0] * (A[1][1] * b[2]    - b[1]    * A[2][1])
                - A[0][1] * (A[1][0] * b[2]    - b[1]    * A[2][0])
                + b[0]    * (A[1][0] * A[2][1] - A[1][1] * A[2][0]);

    x[0] = det0 * inv_det;
    x[1] = det1 * inv_det;
    x[2] = det2 * inv_det;

    return 1;
}


/*
 * Fit y = ax² + bx + c through exactly 3 (x, y) points.
 *
 * pts layout: pts[0] = {x0, y0}, pts[1] = {x1, y1}, pts[2] = {x2, y2}
 * On success, writes [a, b, c] into coeffs and returns coeffs.
 * Returns NULL if the system is degenerate (e.g. two points share the same x).
 * 
 */
double *fit_parabola_3pts(int x[3], int y[3], double coeffs[3]) {
    double x0 = x[0], y0 = y[0];
    double x1 = x[1], y1 = y[1];
    double x2 = x[2], y2 = y[2];

    double A[3][3] = {
        { x0 * x0, x0, 1.0 },
        { x1 * x1, x1, 1.0 },
        { x2 * x2, x2, 1.0 },
    };

    double b[3] = { y0, y1, y2 };

    if (!solve_3x3(A, b, coeffs))
        return NULL; /* mirrors Python's: return None */

    return coeffs;
}


/*
 * Least-squares parabola fit: y = ax² + bx + c
 * Mirrors: np.polyfit(x, y, deg=2)
 *
 * Assumes the following helper is already implemented elsewhere:
 *   int solve_3x3(double A[3][3], double b[3], double x[3]);
 *   returns 1 on success, 0 if singular.
 *
 * Returns 1 on success, 0 on failure.
 * On success, coeffs[0]=a, coeffs[1]=b, coeffs[2]=c.
 * 
 */
int polyfit2(const double *x, const double *y, int n, double coeffs[3])
{
    /* ── accumulate sums ─────────────────────────────────────────────────── */
    double sx1 = 0, sx2 = 0, sx3 = 0, sx4 = 0;
    double sy  = 0, sxy = 0, sx2y = 0;
 
    for (int i = 0; i < n; i++) {
        double xi  = x[i];
        double yi  = y[i];
        double xi2 = xi  * xi;
        double xi3 = xi2 * xi;
        double xi4 = xi3 * xi;
 
        sx1  += xi;
        sx2  += xi2;
        sx3  += xi3;
        sx4  += xi4;
        sy   += yi;
        sxy  += xi  * yi;
        sx2y += xi2 * yi;
    }
 
    /* ── build the normal equations and delegate to solve_3x3 ───────────── */
    double A[3][3] = {
        { sx4, sx3, sx2 },
        { sx3, sx2, sx1 },
        { sx2, sx1, (double)n }
    };
    double b[3] = { sx2y, sxy, sy };
    
    // returns 1 on success, 0 on failure. 
    // The "coeffs" array is modified in-place because it's passed as a pointer.
    return solve_3x3(A, b, coeffs);
}


/*
 * Evaluate the parabola on given x-values.
 *
 * Returns: y, vector the same size as x, rounded to the nearest integer.
 * 
 */
int eval_parabola(const double coeffs[3], int x) {
  double a = coeffs[0];
  double b = coeffs[1];
  double c = coeffs[2];

  // compute result
  double y = a * x*x + b * x + c;
  
  // turn y into an "int" type after rounding
  return (int)round(y);
}


/* 
 * Random sample of 3 distinct indices in [0, n) 
 * Mirrors: np.random.choice(n, size=3, replace=False)
 * Uses Fisher-Yates partial shuffle on a small index buffer.
 *                
 */
static void random_choice_3(int n, int out[3])
{ 
    // create range of numbers [0, n-1]
    int *idx = malloc(n * sizeof(int));
    for (int i = 0; i < n; i++) idx[i] = i;

    for (int i = 0; i < 3; i++) {
        int j = i + rand() % (n - i);   /* random position in [i, n) */
        int tmp = idx[i]; idx[i] = idx[j]; idx[j] = tmp;
        out[i] = idx[i];
    }
    free(idx);
}
 

/* 
 * Fit parabola to a noisy set of points using the RANSAC algorithm.
 *
 */
int ransac_parabola(const int *x, const int *y, int n,
                    int iterations, double threshold, int min_inliers,
                    double best_coeffs[3], int *inlier_mask)
{
	memset(inlier_mask, 0, n * sizeof(int));

	if (n < 3) 
		return 0;

	int  best_inlier_count = 0;
	int *best_mask    = calloc(n, sizeof(int));
	int *current_mask = malloc(n * sizeof(int));

	for (int iter = 0; iter < iterations; iter++) {

		/* ── 3a. RANDOM SAMPLE ──────────────────────────────────────────── */
		int sample[3];
		random_choice_3(n, sample);

		/* ── 3b. FIT MODEL TO SAMPLE ────────────────────────────────────── */
		double  coeffs[3];
		int     x_points[3];
		int     y_points[3];

		// remember, x and y are passed as pointers inside the functionm, 
		// so if I call x[25] it returns the 25th memory slot after the beginning of x.
		// If x and y were passed as values, it wouldn't make sense in C.
		for (int i = 0; i < 3; i++) {
			x_points[i] = x[sample[i]];
			y_points[i] = y[sample[i]];
		}

		if (fit_parabola_3pts(x_points, y_points, coeffs) == NULL)
			continue;   /* degenerate sample, skip */

		/* ── 3c. EVALUATE ALL POINTS AGAINST THIS MODEL ─────────────────── */
		int inlier_count = 0;
		for (int i = 0; i < n; i++) {
				double residual = fabs(eval_parabola(coeffs, x[i]) - y[i]);
				current_mask[i] = (residual < threshold) ? 1 : 0;
				if (current_mask[i]) inlier_count++;
		}

		/* ── 3d. KEEP BEST MODEL ────────────────────────────────────────── */
		// the memcpy function saves its argument in memory.
		if (inlier_count > best_inlier_count) {
				best_inlier_count = inlier_count;
				memcpy(best_coeffs, coeffs, 3 * sizeof(double));
				memcpy(best_mask,   current_mask, n * sizeof(int));
		}
	}

	/* ── 3e. REFINEMENT ─────────────────────────────────────────────────── */
	int success = 0;
	if (best_inlier_count >= min_inliers) {
		double *xi = malloc(best_inlier_count * sizeof(double));
		double *yi = malloc(best_inlier_count * sizeof(double));
		int k = 0;
		for (int i = 0; i < n; i++) {
			// advance k only when an in-lier is copied, so that 
			// xi and yi are tightly packed and not sparse (they're not a mask)
			if (best_mask[i]) { xi[k] = x[i]; yi[k] = y[i]; k++; }
		}

		if (polyfit2(xi, yi, best_inlier_count, best_coeffs)) {
			memcpy(inlier_mask, best_mask, n * sizeof(int));
			success = 1;
		}

		free(xi);
		free(yi);
	} else {
		printf("RANSAC failed: not enough inliers found.\n");
	}

	free(best_mask);
	free(current_mask);
	return success;
}


#include <math.h>
#include "image.h"
/*
 * Draws y = ax² + bx + c onto struct image_t.
 *
 * Assumes the following helpers are already implemented elsewhere:
 *   double eval_parabola(const double coeffs[3], double xval);
 *   void   set_color_yuv422(struct image_t *im, int x, int y, uint8_t Y, uint8_t U, uint8_t V);
 * 
 */
void draw_parabola_on_image(struct image_t *img,
                            const double coeffs[3],
                            uint8_t Y, uint8_t U, uint8_t V,
                            int thickness)
{  
	int thickness = 2;
	int width     = img->h;   /* rows  — as in the original Python */
	int height 		= img->w;   /* cols  — as in the original Python */

	/* Iterate over every row (x in the original Python) */
	for (int x = 0; x < width; x++) {

		/* Compute the column this parabola passes through at this row */
		int y = (int)round(eval_parabola(coeffs, (double)x));

		/* Draw `thickness` pixels horizontally centered on that column */
		for (int dy = -(thickness / 2); dy <= thickness / 2; dy++) {
			int y_draw = y + dy;

			/* Skip if outside image bounds */
			if (y_draw < 0 || y_draw >= height)
					continue;

			// RED in YUV format
			uint8_t Y = 76, U = 84, V = 255;

			set_color_yuv422(img, x, y_draw, Y, U, V);
		}
	}
}

File: DEVELOPMENT/Python/gd_riccardo/ransac_fit.py
import numpy as np

# ── STEP 1: HELPER — FIT PARABOLA THROUGH 3 POINTS ───────────────────────────
# RANSAC needs to fit a model to a minimal sample.
# For a degree-2 polynomial (y = ax² + bx + c), the minimal sample is 3 points.
# We solve the system analytically using numpy — no external solver needed.

def fit_parabola_3pts(pts):
    """
    Fit y = ax² + bx + c through exactly 3 (x, y) points.
    Returns (a, b, c) or None if the points are degenerate (e.g. same x).
    
    We build the Vandermonde matrix:
        | x0²  x0  1 |   | a |   | y0 |
        | x1²  x1  1 | × | b | = | y1 |
        | x2²  x2  1 |   | c |   | y2 |
    and solve it with np.linalg.solve.
    """
    (x0, y0), (x1, y1), (x2, y2) = pts

    # Build the 3×3 Vandermonde matrix
    A = np.array([
        [x0**2, x0, 1],
        [x1**2, x1, 1],
        [x2**2, x2, 1],
    ])

    b = np.array([y0, y1, y2])

    try:
        coeffs = np.linalg.solve(A, b)  # returns [a, b, c]
    except np.linalg.LinAlgError:
        return None  # matrix was singular (e.g. two points share the same x)

    return coeffs  # [a, b, c]


# ── STEP 2: HELPER — EVALUATE PARABOLA ────────────────────────────────────────
# Given coefficients and an array of x values, compute predicted y values.

def eval_parabola(coeffs, x):
    """
    Evaluate y = ax² + bx + c at position(s) x.
    coeffs: array [a, b, c]
    x: scalar or numpy array
    """
    a, b, c = coeffs
    return a * x**2 + b * x + c


# ── STEP 3: RANSAC MAIN LOOP ──────────────────────────────────────────────────
# RANSAC works by repeatedly:
#   1. Picking a tiny random sample (3 points for a parabola)
#   2. Fitting the model to that sample
#   3. Counting how many OTHER points are close to that model (inliers)
#   4. Keeping the model that has the most inliers
#
# The key insight: even if 30-40% of your data are outliers,
# there's a good chance that some random draws of 3 points will
# pick 3 clean inliers and produce a good fit.

def ransac_parabola(x, y, iterations=200, threshold=10.0, min_inliers=50):
    """
    Fit y = ax² + bx + c to (x, y) data robustly using RANSAC.

    Parameters:
        x, y        : 1D arrays of contour points (one per image column)
        iterations  : how many random samples to try
        threshold   : max residual (pixels) to count a point as an inlier
        min_inliers : minimum inliers required to accept a fit

    Returns:
        best_coeffs : array [a, b, c] of the best fit found, or None
        inlier_mask : boolean array marking which points are inliers
    """
    n = len(x)
    best_coeffs = None
    best_inlier_count = 0
    best_inlier_mask = np.zeros(n, dtype=bool)

    if n < 3:
        return None, np.zeros(n, dtype=bool)

    for _ in range(iterations):

        # ── 3a. RANDOM SAMPLE ─────────────────────────────────────────────────
        # Pick 3 distinct random indices from the data.
        # These 3 points fully determine a parabola.
        sample_idx = np.random.choice(n, size=3, replace=False)
        sample_pts = [(x[idx], y[idx]) for idx in sample_idx]

        # ── 3b. FIT MODEL TO SAMPLE ───────────────────────────────────────────
        # Fit a parabola through just these 3 points.
        coeffs = fit_parabola_3pts(sample_pts)
        if coeffs is None:
            continue  # degenerate sample, skip

        # ── 3c. EVALUATE ALL POINTS AGAINST THIS MODEL ────────────────────────
        # For every point in the dataset, compute how far it is
        # from the candidate parabola (the "residual").
        y_predicted = eval_parabola(coeffs, x)
        residuals = np.abs(y_predicted - y)

        # A point is an "inlier" if its residual is within the threshold.
        # Inliers are consistent with this candidate model.
        # Outliers (obstacles, plants) will have large residuals.
        inlier_mask = residuals < threshold
        inlier_count = inlier_mask.sum()

        # ── 3d. KEEP BEST MODEL ───────────────────────────────────────────────
        # If this candidate has more inliers than any previous one, save it.
        if inlier_count > best_inlier_count:
            best_inlier_count = inlier_count
            best_coeffs = coeffs
            best_inlier_mask = inlier_mask

    # ── 3e. OPTIONAL REFINEMENT ───────────────────────────────────────────────
    # Once we know which points are inliers, we can refit the parabola
    # using ALL inliers (not just the 3 random points), which gives a
    # more accurate final result. This is called the "refinement step".
    if best_coeffs is not None and best_inlier_count >= min_inliers:
        x_inliers = x[best_inlier_mask]
        y_inliers = y[best_inlier_mask]
        # np.polyfit does a least-squares fit — safe here because
        # we've already removed the outliers via RANSAC
        best_coeffs = np.polyfit(x_inliers, y_inliers, deg=2)
    else:
        print("RANSAC failed: not enough inliers found.")
        best_coeffs = None

    return best_coeffs, best_inlier_mask


def draw_parabola_on_image(image_bgr, coeffs, x, y, color=(0, 0, 255), thickness=2):

    width, height = image_bgr.shape[:2]
    x_all = np.arange(width)

    # Compute all y values at once
    y_all = np.round(eval_parabola(coeffs, x_all)).astype(int)

    # x_all = x.astype(bool)
    # y_all = y.astype(bool)

    # For each thickness offset
    for dy in range(-thickness // 2, thickness // 2 + 1):
        y_draw = y_all + dy

        # Only keep columns where y_draw is within the image
        valid = (y_draw >= 0) & (y_draw < height)

        # Index directly: image[row, col] = color
        image_bgr[x_all[valid], y_draw[valid]] = color

    return image_bgr



# code inside main.py
if __name__ == "__main__":

    pass